src/Neo/Ledger/Blockchain.cs

```diff
@@ -414,26 +416,47 @@ private void Persist(Block block)
                     all_application_executed.Add(application_executed);
                     transactionStates = engine.GetState<TransactionState[]>();
                 }
-                DataCache clonedSnapshot = snapshot.CreateSnapshot();
-                // Warning: Do not write into variable snapshot directly. Write into variable clonedSnapshot and commit instead.
+
+                var tasks = new List<Task<(TransactionState, DataCache, ApplicationExecuted)>>();
+
+                // Warning: Do not write into variable snapshot directly.
+                // Write into variable clonedSnapshot and commit instead.
                 foreach (TransactionState transactionState in transactionStates)
                 {
-                    Transaction tx = transactionState.Transaction;
-                    using ApplicationEngine engine = ApplicationEngine.Create(TriggerType.Application, tx, clonedSnapshot, block, system.Settings, tx.SystemFee);
-                    engine.LoadScript(tx.Script);
-                    transactionState.State = engine.Execute();
-                    if (transactionState.State == VMState.HALT)
+                    DataCache txSnapshot = snapshot.CreateSnapshot();
+                    var task = OnExecuteTransactionAsync(system, block, txSnapshot, transactionState);
+                    tasks.Add(task);
+                }
+
+                // var readSet = new HashSet<StorageKey>();
+                var writeSet = new HashSet<StorageKey>();
+
+                var randomUsed = false;
+                foreach (var task in tasks)
+                {
+                    var (transactionState, txSnapshot, executed) = task.Result;
+
+                    if (txSnapshot.GetReadSet().Overlaps(writeSet) || (randomUsed && txSnapshot.isRandomNumberCalled))
                     {
-                        clonedSnapshot.Commit();
+                        DataCache tmp = snapshot.CreateSnapshot();
+                        var task2 = OnExecuteTransactionAsync(system, block, tmp, transactionState);
+                        tasks.Add(task2);
```

Comment:
Why adding to the list if you're waiting for the result below anyway?

Reply:
parallel execution, sequential  commit.

In this way, we can start processing commit in the original order on txs that are alrady finished  without waiting for the whole execution to finish.